source/tests/system/nidigital_driver_api_tests.cpp

```diff
@@ -0,0 +1,292 @@
+#include <gtest/gtest.h>
+
+#include "nidigital/nidigital_library.h"
+#include "nidigital/nidigital_service.h"
+
+namespace ni {
+namespace tests {
+namespace system {
+
+namespace digital = nidigital_grpc;
+
+const int kDigitalDriverApiSuccess = 0;
+
+class NiDigitalDriverApiTest : public ::testing::Test {
+ protected:
+  NiDigitalDriverApiTest()
+  {
+    ::grpc::ServerBuilder builder;
+    session_repository_ = std::make_unique<nidevice_grpc::SessionRepository>();
+    nidigital_library_ = std::make_unique<digital::NiDigitalLibrary>();
+    nidigital_service_ = std::make_unique<digital::NiDigitalService>(nidigital_library_.get(), session_repository_.get());
+    builder.RegisterService(nidigital_service_.get());
+
+    server_ = builder.BuildAndStart();
+    ResetStub();
+  }
+
+  virtual ~NiDigitalDriverApiTest() {}
+
+  void SetUp() override
+  {
+    initialize_driver_session();
+  }
+
+  void TearDown() override
+  {
+    close_driver_session();
+  }
+
+  void ResetStub()
+  {
+    channel_ = server_->InProcessChannel(::grpc::ChannelArguments());
+    nidigital_stub_ = digital::NiDigital::NewStub(channel_);
+  }
+
+  std::unique_ptr<digital::NiDigital::Stub>& GetStub()
+  {
+    return nidigital_stub_;
+  }
+
+  int GetSessionId()
+  {
+    return driver_session_->id();
+  }
+
+  void initialize_driver_session()
+  {
+    ::grpc::ClientContext context;
+    digital::InitWithOptionsRequest request;
+    request.set_resource_name("FakeDevice");
+    request.set_option_string("Simulate=1, DriverSetup=Model:6570; BoardType:PXIe");
+    request.set_session_name("");
+    request.set_reset_device(false);
+    request.set_id_query(false);
+    digital::InitWithOptionsResponse response;
+
+    ::grpc::Status status = GetStub()->InitWithOptions(&context, request, &response);
+    driver_session_ = std::make_unique<nidevice_grpc::Session>(response.vi());
+
+    ASSERT_TRUE(status.ok());
+    ASSERT_EQ(kDigitalDriverApiSuccess, response.status());
+  }
+
+  void close_driver_session()
+  {
+    ::grpc::ClientContext context;
+    digital::CloseRequest request;
+    request.mutable_vi()->set_id(driver_session_->id());
+    digital::CloseResponse response;
+
+    ::grpc::Status status = GetStub()->Close(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+  }
+
+  void expect_api_success(int error_status)
+  {
+    EXPECT_EQ(kDigitalDriverApiSuccess, error_status) << get_error_message(error_status);
+  }
+
+  std::string get_error_message(int error_status)
+  {
+    ::grpc::ClientContext context;
+    digital::ErrorMessageRequest request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_error_code(error_status);
+    digital::ErrorMessageResponse response;
+
+    ::grpc::Status status = GetStub()->ErrorMessage(&context, request, &response);
```

Comment:
The newline should be after the library call - to separate out "action" and "validation" part of the method.